docs/keps/20200619-kubefed-pull-reconciliation.md

```diff
@@ -0,0 +1,314 @@
+
+This first iteration provides a decentralized reconciliation and propagation where the daemons directly talk to
+the Kubernetes API server to ensure the desired state.
+
+However this iteration does not filter which federated resources can be viewed or accessed by each kubefed cluster. This would be addressed as
+part of the second iteration.
+
+
+#### Second iteration
+
+During the **second iteration**, the kubefed control-plane deploys an additional component.
+This component works as a shim API server to which the kubefed daemons talk to perform
+all the operations done via the Kubernetes API server in the **first iteration**.
+The shim API server limits what each daemon can view or access and reduces the load of the daemons on the Kubernetes API server.
+
+The main challenge of this shim API server is filtering which federated resources a kubefed
+daemon should be able to `get`, `watch`, `list` and `update` their status.
+
+This server watches for changes in the federated resources, and does a fan-out to it's daemons.
+This also reduces the load on the Kubernetes API server significantly cause kubefed daemons are talking to this server.
+All the kubefed daemon operations on federated resources are handled by the shim API server that works **similarly** to an authorizers and an aggregated API combined.
+A daemon should be only aware of changes in the resources assigned to its own cluster.
+
+Without this new control-plane component, every kubefed daemon would have to perform any operations and register its own watch via the Kubernetes API Server.
+Consequently the load on API server would multiply as you scale up the number of kubefed daemons, and the system would lack of a proper secure layer to limit what each daemon can access or do.
+
+<img src="./images/kubefedv2_seconditeration.jpg">
+
+By having this API server, all the watch events are off-loaded to this component which filters the visibility of resources from the Kubernetes API server.
+
+#### Other Main Reconciliation Loops
+
+As part of this new reconciliation pull mode, there are certain controllers that
+could become deprecated or removed.
+
+For the `ServiceDNSRecord` controller, Kubefed `v2` should omit this functionality and rely on
+other solutions such as service meshes.
+The purpose of Kubefed should be to ultimately federate resources across clusters.
+Another option is to expose an additional endpoint in the daemons to be aware of the
+`serviceDNS`, therefore the control-plane does not need to create remote `Informer` against
+all the kubefed clusters.
+
+The future of `ServiceDNSRecord` controller is something to decide with the community.
+It goes one step beyond by dealing with networking considerations when talking to apps deployed across clusters.
+
+On the other hand, the controllers in charge of the scheduler preferences remains as part of the control-plane.
```

Comment:
I feel that we should deprecate and remove scheduling preference api types from this project in v2 and keep this project focused on resource federation. The scheduling preference problem can be solved in a higher level leveraging primitives provided by this project.